primitives/trie/src/storage_proof.rs

```diff
@@ -14,65 +14,990 @@
+							match children.entry(child_info) {
+								btree_map::Entry::Occupied(v) => if v.get() != &root {
+									return false;
+								},
+								btree_map::Entry::Vacant(v) => {
+									v.insert(root);
+								},
+							}
+						}
+					},
+					Input::QueryPlan(..) => return false,
+					Input::QueryPlanWithValues(..) => return false,
+				}
+			},
+			Input::QueryPlan(..) => return false,
+			Input::QueryPlanWithValues(..) => return false,
+		}
+		true
+	}
+}
+
+/// Kind for a `Input` variant.
+pub enum InputKind {
+	/// `Input::None` kind.
+	None,
+
+	/// `Input::ChildTrieRoots` kind.
+	ChildTrieRoots,
+
+	/// `Input::QueryPlan` kind.
+	QueryPlan,
+
+	/// `Input::QueryPlanWithValues` kind.
+	QueryPlanWithValues,
+}
+
+impl StorageProofKind {
+	/// Input kind needed for processing (create) the proof.
+	pub fn process_input_kind(&self) -> InputKind {
+		match self {
+			StorageProofKind::KnownQueryPlanAndValues => InputKind::QueryPlan,
+			StorageProofKind::TrieSkipHashesForMerge
+				| StorageProofKind::TrieSkipHashes
+				| StorageProofKind::TrieSkipHashesFull => InputKind::ChildTrieRoots,
+			StorageProofKind::Full
+				| StorageProofKind::Flatten => InputKind::None,
+		}
+	}
+
+	/// Input kind needed for verifying the proof.
+	pub fn verify_input_kind(&self) -> InputKind {
+		match self {
+			StorageProofKind::KnownQueryPlanAndValues => InputKind::QueryPlanWithValues,
+			StorageProofKind::TrieSkipHashes
+				| StorageProofKind::TrieSkipHashesFull
+				| StorageProofKind::TrieSkipHashesForMerge
+				| StorageProofKind::Full
+				| StorageProofKind::Flatten => InputKind::None,
+		}
+	}
+
+	/// Indicates what variant of proof recorder should be use.
+	pub fn need_register_full(&self) -> bool {
+		match self {
+			StorageProofKind::Flatten => false,
+			StorageProofKind::Full
+				| StorageProofKind::KnownQueryPlanAndValues
+				| StorageProofKind::TrieSkipHashes
+				| StorageProofKind::TrieSkipHashesForMerge
+				| StorageProofKind::TrieSkipHashesFull => true,
+		}
+	}
+
+	/// Indicates if we should execute proof over a backend,
+	/// and if so, if the backend need to be full.
+	pub fn use_full_partial_db(&self) -> Option<bool> {
+		match self {
+			StorageProofKind::Flatten
+				| StorageProofKind::TrieSkipHashes => Some(false),
+			StorageProofKind::Full
+				| StorageProofKind::TrieSkipHashesForMerge
+				| StorageProofKind::TrieSkipHashesFull => Some(true),
+			StorageProofKind::KnownQueryPlanAndValues => None,
+		}
+	}
+
+	/// Proof that should be use with `verify` method.
+	pub fn can_use_verify(&self) -> bool {
+		matches!(self.verify_input_kind(), InputKind::None)
+	}
+
+	/// Can be use as a trie db backend.
+	pub fn can_use_as_partial_db(&self) -> bool {
+		match self {
+			StorageProofKind::KnownQueryPlanAndValues => false,
+			_ => true,
+		}
+	}
+
+	/// Return the best kind to use for merging later,
+	/// a boolean indicationg if merge should produce full proof.
+	pub fn mergeable_kind(&self) -> (Self, bool) {
```

Comment:
I feel like all methods in this type strongly suggest that the proof kinds should rather be separate types where we can rely on the typesystem to make sure they are used in correct way instead of doing a bunch of runtime checks for `input_kind` or mergeability etc.